deepchem/models/normalizing_flows.py

```diff
@@ -0,0 +1,401 @@
+
+  A distribution implements two main operations:
+    1. Sampling from the transformed distribution.
+    2. Calculating log probabilities.
+
+  A normalizing flow implements three main operations:
+    1. Forward transformation, 2. Inverse transformation, and 
+    3. Calculating the Jacobian.
+
+  Deep Normalizing Flow models require normalizing flow layers where
+  input and output dimensions are the same, the transformation is invertible,
+  and the determinant of the Jacobian is efficient to compute and
+  differentiable. The determinant of the Jacobian of the transformation 
+  gives a factor that preserves the probability volume to 1 when transforming
+  between probability densities of different random variables.
+
+  They are effective for any application requiring a probabilistic
+  model with these capabilities, e.g. generative modeling,
+  unsupervised learning, or probabilistic inference. For a thorough review
+  of normalizing flows, see [1]_.
+
+  References
+  ----------
+  .. [1] Papamakarios, George et al. "Normalizing Flows for Probabilistic Modeling and Inference." (2019). https://arxiv.org/abs/1912.02762.
+
+  """
+
+  def __init__(self,
+               model: NormalizingFlow,
+               loss=NegLogLoss,
+               optimizer=Adam,
+               learning_rate: float = 1e-5,
+               batch_size: int = 64,
+               **kwargs):
+    """Creates a new NormalizingFlowModel.
+
+    Parameters
+    ----------
+    model: NormalizingFlow
+      An instance of NormalizingFlow.
+    loss: dc.models.losses.Loss, default NegLogLoss
+      Loss function
+    optimizer: dc.models.optimizers.Optimizer, default Adam
+      Optimizer.
+    learning_rate: float, default 1e-5
+      Learning rate for optimizer.
+    
+
+    Examples
+    --------
+    >> import tensorflow_probability as tfp
+    >> tfd = tfp.distributions
+    >> tfb = tfp.bijectors
+    >> flow_layers = [
+    ..    tfb.RealNVP(
+    ..        num_masked=2,
+    ..        shift_and_log_scale_fn=tfb.real_nvp_default_template(
+    ..            hidden_layers=[8, 8]))
+    ..]
+    >> base_distribution = tfd.MultivariateNormalDiag(loc=[0., 0., 0.])
+    >> nf = NormalizingFlow(base_distribution, flow_layers)
+    >> nfm = NormalizingFlowModel(nf)
+    >> dataset = NumpyDataset(
+    ..    X=np.random.rand(5, 3).astype(np.float32),
+    ..    y=np.random.rand(5,),
+    ..    ids=np.arange(5))
+    >> nfm.fit(dataset)
+
+    """
+
+    try:
+      import tensorflow_probability as tfp
+      tfd = tfp.distributions
+      tfb = tfp.bijectors
+    except ModuleNotFoundError:
+      raise ValueError(
+          "This class requires tensorflow-probability to be installed.")
+
+    self.model = model
+    self.flow = model.flow  # normalizing flow
+    self.loss = loss()
+    self.batch_size = batch_size
+    self.learning_rate = learning_rate
+
+    self.optimizer = optimizer(learning_rate=learning_rate)
+
+    self.built = False
+    self.build()
```

Comment:
Most of this conflicts with what the superclass is already doing.  For model, batch_size, learning_rate, etc. you don't even need to explicitly list arguments.  KerasModel will pull them out of `kwargs` and set appropriate fields.  Take a look at `__init__()` and `_ensure_built()` in KerasModel.